res/res_format_attr_red.c

```diff
@@ -0,0 +1,283 @@
+{
+	struct red_attr *attr = ast_format_get_attribute_data(format);
+	int base_fmtp_size;
+	int original_size;
+	if (!attr) {
+		ast_log(LOG_ERROR, "Invalid RED attributes\n");
+		return;
+	}
+
+	/*
+	* Generate the SDP fmtp line for RED
+	* Assuming attr->red_payload holds the payload types for redundancy
+	* and attr->red_num_gen holds the number of generations
+	*/
+	original_size = ast_str_strlen(*str);
+	base_fmtp_size = ast_str_append(str, 0, "a=fmtp:%u ", payload);
+
+	if (attr->red_num_gen > 0) {
+		/* attributes are one more as the number of generations */
+		for (int i = 0; i < attr->red_num_gen + 1; i++) {
+			ast_str_append(str, 0, "%d/", attr->red_payload);
+		}
+	}
+
+	if (base_fmtp_size == ast_str_strlen(*str) - original_size) {
+		ast_str_truncate(*str, original_size);
+	} else {
+		ast_str_truncate(*str, -1);
+		ast_str_append(str, 0, "\r\n");
+	}
+
+	ast_debug(3, "RED sdp written: %s\n", ast_str_buffer(*str));
+}
+
+
+static const void *red_attribute_get(const struct ast_format *format, const char *name)
+{
+	struct red_attr *attr = ast_format_get_attribute_data(format);
+	int *val;
+
+	if (!strcasecmp(name, "red_num_gen")) {
+		val = &attr->red_num_gen;
+	} else if (!strcasecmp(name, "red_payload")) {
+		val = &attr->red_payload;
+	} else if (!strcasecmp(name, "cps")) {
+		val = &attr->cps;
+	} else {
+		ast_log(LOG_WARNING, "unknown attribute type %s\n", name);
+		return NULL;
+	}
+
+	return val;
+}
+
+
+static struct ast_format *red_attribute_set(const struct ast_format *format, const char *name, const char *value)
+{
+	struct ast_format *cloned;
+	struct red_attr *attr;
+	unsigned int val;
+
+	if (sscanf(value, "%30u", &val) != 1) {
+		ast_log(LOG_WARNING, "Unknown value '%s' for attribute type '%s'\n",
+			value, name);
+		return NULL;
+	}
+
+	cloned = ast_format_clone(format);
+	if (!cloned) {
+		return NULL;
+	}
+	attr = ast_format_get_attribute_data(cloned);
+
+	if (!strcasecmp(name, "red_num_gen")) {
+		attr->red_num_gen = val;
+	} else if (!strcasecmp(name, "red_payload")) {
+		attr->red_payload = val;
+	} else if (!strcasecmp(name, "cps")) {
+		attr->cps = val;
+	} else {
+		ast_log(LOG_WARNING, "unknown attribute type %s\n", name);
+	}
+
+	return cloned;
+}
+
+static struct ast_format *red_getjoint(const struct ast_format *format1, const struct ast_format *format2)
+{
+	struct red_attr *attr1 = ast_format_get_attribute_data(format1);
+	struct red_attr *attr2 = ast_format_get_attribute_data(format2);
+	struct ast_format *jointformat;
+	struct red_attr *attr_res;
+	/* To re-use offered RED payload types */
+	unsigned int use_side = 0;
+
+	if (!attr1) {
+		attr1 = &default_red_attr;
+		use_side = 2;
+	}
+
+	if (!attr2) {
+		attr2 = &default_red_attr;
+		use_side = 1;
+	}
+
+	jointformat = (use_side == 2) ? ast_format_clone(format2) : ast_format_clone(format1);
+	if (!jointformat) {
+		return NULL;
+	}
+
+	// Get the resulting attributes for the joint format
```

Comment:
Don't use `//` style comments. There are multiple occurrences of these but I am only commenting on one.

Reply:
Fixed